readme.md
# Slick Personal Assistant :robot:

![440x280.png](https://raw.githubusercontent.com/spencerslickremix/slickGPT/main/screenshots/440x280.png)


## :book: Description

A Chrome extension that adds [ChatGPT](https://chat.openai.com) to your browser. It can assist in creating content or code for any website based on your suggestions.

Pin the Extension to your browser so you can easily use on any tab. It will work independently between tabs too.

### **Options**
- GPT-3 Turbo or GPT-4 
- Custom Prompts or GitHub Prompts
- Number of Tokens
- Alternative Completions
- Temperature
- Twitter Share Suggestion Icon
- Light and Dark Mode

Below are some use cases and also a breakdown of the Custom Options, Settings and their meaning.

---

## :bulb: Use Cases
1. Say you are editing your website and want to modify some text, or create a blog post. It would be easier to select that text on your website, have it appear automatically in the popup, and then ask ChatGPT to make the changes you want after the STOP. Once a suggestion(s) appears, click it and the selected text on the page will be replaced. By adjusting the Temperature from the Settings tab, you can have the response be less or more creative.
2. You don't have to use this in conjunction with a website. You can also simply use it like ChatGPT for having a conversation and creating or modifying code.
3. Great for writing a creative Tweet that you can easily share on Twitter once you are done.
4. Useful for creative writing in an email or text message. All you have to do is click the response when you are done, and the text is copied to your clipboard.
5. Quickly Translate Text.

---

## :wrench: Custom Options
1. Select text on the web page and click the pinned extension; the selected text will automatically be pasted to the textarea in the chrome extension popup.
2. I added a STOP: after the pasted text so you can easily prompt your request, or not.
3. Once you click the "Submit" button, one or more suggestions will appear.
4. You can then click on a suggestion, and the text from that suggestion will be copied in place of the selected text on the webpage if it's a content editable area.
5. The text will also be added to the clipboard, allowing you to paste it in another tab or anywhere.
6. **Bonus:** Click the Twitter icon that appears in the top right of a suggestion(s), and it will open Twitter so you can share the suggestion.

---

## :pencil: Custom Prompts
1. Add up to 5 custom Descriptions and Prompts.
2. Connect a GitHub account to display as many Prompts as you want. See the [ChatGPT Custom Prompts](https://github.com/spencerslickremix/chatGPT-Prompts) and learn how to connect them to the Chrome Extension. 

---


## :gear: Settings
1. **[API Key](https://platform.openai.com/account/api-keys):** Your OpenAI API key is required to access ChatGPT. Click the API Key link to get yours. You must have an account setup. OpenAI gives you $18 free to start; after that, you will need to add a payment method to continue use. Cheaper than paying $20 a month for ChatGPT Pro.
2. **Max Tokens:** The maximum number of tokens (words or word pieces) in the generated text. 4097 is the maximum allowed per prompt.
3. **Alternative Completions:** The number of alternative completions to generate for each prompt.
4. **Temperature:** Controls the randomness of the generated text. Higher values (e.g., 1.0) result in more creative output, while lower values (e.g., 0.1) produce more coherent text.
5. **Type:** GPT-4 or GPT-3
6. **Share:** Twitter Share option for Suggestions.
7  **Mode:** Light or Dark

---

## :camera: Screenshots

![Screenshot 11.png](https://raw.githubusercontent.com/spencerslickremix/Slick-Personal-Assistant/main/screenshots/Screenshot%206.png)

![Screenshot 1.png](https://raw.githubusercontent.com/spencerslickremix/Slick-Personal-Assistant/main/screenshots/Screenshot%201.png)

![Screenshot 2.png](https://raw.githubusercontent.com/spencerslickremix/Slick-Personal-Assistant/main/screenshots/Screenshot%202.png)

![Screenshot 8.png](https://raw.githubusercontent.com/spencerslickremix/Slick-Personal-Assistant/main/screenshots/Screenshot%203.png)

![Screenshot 9.png](https://raw.githubusercontent.com/spencerslickremix/Slick-Personal-Assistant/main/screenshots/Screenshot%204.png)

![Screenshot 10.png](https://raw.githubusercontent.com/spencerslickremix/Slick-Personal-Assistant/main/screenshots/Screenshot%205.png)

---

## :rocket: Install
Simply click the green CODE button near the top right of this repo. Download the ZIP file and unpack it wherever you want on your computer.

---

### Add the extension

1. Go to `chrome://extensions` in your Google Chrome browser
2. Check the Developer mode checkbox in the top right-hand corner
3. Click "Load Unpacked" to see a file-selection dialog
4. Select the location where the Slick-Personal-Assistant folder is.
5. Click the Google Extensions icon in your browser tab, looks like a puzzle piece.
6. Pin the "SlickRemix: Your ChatGPT Web Assistant" extension to your browser.
7. Add your own API key by clicking the Settings button.

It's that simple, have fun and enjoy!

---

## :scroll: License
2023 MIT © Spencer Labadie . Follow me on <a href="https://twitter.com/SpencerLabadie"><img src="https://img.icons8.com/color/48/000000/twitter.png" width="28"></a>


changelog.txt
= Version 0.0.3 Monday, May 29th, 2023 =
 * New: Settings: Custom font size option.
 * New: Settings: Custom background and color options.
 * New: Settings: Choose the width of the Popup.
 * New: Suggestions: Added Token Count so users can see how many tokens are used per response. This text will also link you directly to the Settings page for quicker navigation.
 * New: Header: The SlickRemix logo now links to the website, and the text "Your Personal Automated Assistant" links to the Suggestions tab for quicker navigation.
 * New: Organized files in their respective folders.
 * New: Namespaced and classify entire extension.
 * Updated: Select a portion of text from a response to be copied automatically.
 * Updated: Settings description for Max Tokens and Temperature to be more clear.
 * Fixed: Added additional checks to Regex to parse URLs properly. We will probably end up adding a library in the future.
 * Fixed: Added error messages to the suggestion list so users know when there is a problem.
 * Fixed: Added GPT-4 specific error message to let users know if the API token they have entered will work with GPT-4 if they chose that model.
 * Removed: content_scripts: matches": ["http://*/*", "https://*/*"] & js": ["contentScript.js"]. These options were not needed because I am calling chrome.scripting.executeScript from the popup.js file already, so the call was being duplicated.

= Version 0.0.2 Thursday, May 18th, 2023 =
 * New: Replaced model davinci-003 with gpt-3.5-turbo
 * New: Updated completions to use /chat/completions
 * New: Save user and response text to array to chrome.storage
 * New: Complete UI change. Made Dark Mode true by default now.
 * New: Regex to parse URLs properly.
 * New: Settings: Display Twitter icon to quickly share responses.
 * New: Custom Prompts tab. Includes the option to add up to 5 custom prompts or chose to display prompts saved from a GitHub accounts readme.md file.
 * New: Showdown.js to convert code to HTML and Prism.js to beatify the code.
 * Removed: permissions: tabs. Was not actually needed and was warned by Google it needed to be removed.

= Version 0.0.1 Monday, May 2nd, 2023 =
 * Initial Release

background.js
// Define your token count at the top of the script
let tokenCount = 0;

// 4/25/23: Need to revisit this and dig deeper into seeing
// if we are actually removing suggestion to be able to keep the conversation going.
// at some point the conversation cannot continue if it exceeds 4097 tokens.
// seems like I need to make an option to save the conversation and or something...
// Define the token counting function
function countTokens(text) {
    let wordCount = text.split(' ').length;
    let punctuationCount = (text.match(/[.,\/#!$%\^&\*;:{}=\-_`~()]/g) || []).length;
    return wordCount + punctuationCount;
}

async function fetchApiKey() {
    const { apiKey } = await chrome.storage.sync.get("apiKey");
    return apiKey;
}

// Initialize the conversation in storage
chrome.runtime.onInstalled.addListener(() => {
    chrome.storage.sync.set({ conversation: [] });
});

// Only use this to test the notification notice.
// The version should be one less that the version in the manifest file.
/*chrome.storage.sync.set({ lastNotifiedVersion: '0.0.2', notificationClicked: 'no' }, () => {
    console.log('Version set in storage and notification option stored');
});*/
/*chrome.storage.sync.get("lastNotifiedVersion", function(data) {
    console.log('Last notified version:', data.lastNotifiedVersion);
});
chrome.storage.sync.get("notificationClicked", function(data) {
    console.log('Was notification clicked:', data.notificationClicked);
});*/

// Update the last notified version in storage when the extension is updated
chrome.runtime.onInstalled.addListener(details => {
    if (details.reason === "update") {
        const currentVersion = chrome.runtime.getManifest().version;
        chrome.storage.sync.set({ lastNotifiedVersion: currentVersion });
    }
});

// Update the conversation in the storage whenever it changes
function updateConversation(convo) {
    chrome.storage.sync.set({ conversation: convo });
}

let conversation = [];
chrome.storage.sync.get("conversation", function(data) {
    if (data.conversation !== undefined) {
        conversation = data.conversation;
        tokenCount = conversation.reduce((sum, message) => sum + message.tokens, 0);
    }
});

async function getSuggestionsFromApi(apiKey, prompt, maxTokens, n, stop, temperature, engine) {

    if (!apiKey) {
        throw new Error("API key not provided. To get started please click on the Menu option in the top right corner, click on Settings and then enter your API key.");
    }
    // gpt-3.5-turbo
    const model = engine === 'gpt3' ? 'gpt-3.5-turbo' : 'gpt-4';

    const engineURL = 'https://api.openai.com/v1/chat/completions';

    let promptTokens = countTokens(prompt);
    conversation.push({role: "user", content: prompt, tokens: promptTokens});
    tokenCount += promptTokens;

    while (tokenCount > 4096 && conversation.length > 0) {
        let removedMessage = conversation.shift();
        tokenCount -= removedMessage.tokens;  // Subtract the tokens of the removed message
    }

    if (promptTokens > 4096) {
        throw new Error("Your message is too long! Please limit messages to 4096 tokens.");
    }

    // copy the conversation array and remove tokens property
    const conversationCopy = conversation.map(({ role, content }) => ({ role, content }));

    const response = await fetch(engineURL, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
            "Authorization": `Bearer ${apiKey}`
        },
        body: JSON.stringify({
            model: model,
            messages: conversationCopy,
            max_tokens: parseInt(maxTokens) || 2000,
            n: parseInt(n) || 1,
            stop: stop || null,
            temperature: parseFloat(temperature) || 0.1
        })
    });

    if (!response.ok) {
        const errorData = await response.json();
        if (errorData.error && errorData.error.code === "invalid_api_key") {
            throw new Error("Invalid API key provided. Please check your API key.");
        }
        else if (errorData.error && errorData.error.message === "The model: `gpt-4` does not exist") {
            //console.log( errorData.error );
            throw new Error("Please choose the GPT-3 Option from the <span class='settingsLink'>Settings</span> page to continue. The OpenAI API Key you entered does not have GPT-4 access yet. <br/><br/>You must have explicitly been granted a GPT-4 API Key from OpenAI for the GPT-4 option to work on our Settings page. You can apply for access here https://openai.com/waitlist/gpt-4-api");
        }
        else{
            throw new Error("Unexpected response from OpenAI API: " + JSON.stringify(errorData));
        }
    }

    const data = await response.json();
    if (data.choices) {
        let suggestionResponses = data.choices.map((choice) => {
            let messageContent = choice.message.content.trim();
            let tokens = countTokens(messageContent);
            conversation.push({role: "assistant", content: messageContent, tokens: tokens});

            // Notify loading-suggestions.js that a new suggestion has been added
            chrome.runtime.sendMessage({ action: "newSuggestionAdded", tokens: tokens });

            return messageContent;
        });
        return suggestionResponses;
    } else {
        console.error("Unexpected response from OpenAI API:", JSON.stringify(data, null, 2));
        return [];
    }
}


chrome.runtime.onMessage.addListener((request, sender, sendResponse) => {
    if (request.action === "getSuggestions") {
        if (request.text) {
            fetchApiKey().then(async function (apiKey) {
                const settings = await chrome.storage.sync.get(["prompt", "maxTokens", "n", "stop", "temperature", "engine"]);
                try {
                    const suggestions = await getSuggestionsFromApi(
                        apiKey,
                        request.text,
                        settings.maxTokens || 2000,
                        settings.n || 1,
                        settings.stop || null,
                        settings.temperature || 0.1,
                        settings.engine || 'gpt3'
                    );
                    tokenCount += suggestions.reduce((sum, suggestion) => sum + countTokens(suggestion), 0);
                    sendResponse({ suggestions: suggestions });
                } catch (error) {
                    console.error('Error:', error);
                    sendResponse({ error: error.message });
                }
            }).catch(function (error) {
                console.error("Error fetching API key or settings:", error);
            });
            return true; // Required for async sendResponse
        } else {
            console.error("No text provided.");
        }
    }

    if (request.action === "clearConversationAndTokenCount") {
        conversation = [];  // Clear the conversation array
        tokenCount = 0;     // Clear the token count
        updateConversation(conversation);  // Save the cleared conversation
        sendResponse({ message: "Conversation and token count cleared." });
        return true;
    }



    if (request.action === "fetchChangelog") {
        fetchChangelog()
            .then(changelog => sendResponse({changelog: changelog}))
            .catch(error => sendResponse({error: error.message}));
        return true;  // Indicate that we will send a response asynchronously
    }
});

function fetchChangelog(attempt = 1) {
    const maxAttempts = 3;
    const repo = 'https://raw.githubusercontent.com/spencerslickremix/Slick-Personal-Assistant/main/';

    return new Promise((resolve, reject) => {
        fetch(repo + 'changelog.txt')
            .then(response => {
                if (response.status === 404) {
                    throw new Error('The changelog could not be found. Please check the file path.');
                } else if (!response.ok) {
                    throw new Error(`HTTP error! status: ${response.status}`);
                }
                return response.text();
            })
            .then(changelog => {
                resolve(changelog);
            })
            .catch(error => {
                if (error.message.includes('changelog could not be found')) {
                    // If the changelog was not found, reject the promise immediately
                    reject(error);
                } else if (attempt < maxAttempts) {
                    // If the fetch failed for a different reason and we haven't reached the max number of attempts, try again
                    fetchChangelog(attempt + 1).then(resolve).catch(reject);
                } else {
                    // If we've reached the max number of attempts, reject the promise
                    reject(error);
                }
            });
    });
}
